src/Storages/StorageTimeSeries.cpp

```diff
@@ -0,0 +1,524 @@
+#include <Storages/StorageTimeSeries.h>
+
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <DataTypes/DataTypeMap.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/DatabaseCatalog.h>
+#include <Interpreters/InterpreterCreateQuery.h>
+#include <Interpreters/InterpreterDropQuery.h>
+#include <Parsers/ASTDropQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTTargetTables.h>
+#include <Storages/AlterCommands.h>
+#include <Storages/StorageFactory.h>
+#include <base/insertAtEnd.h>
+#include <filesystem>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_QUERY;
+}
+
+
+namespace
+{
+    namespace fs = std::filesystem;
+
+    const ColumnsDescription & getColumnsDescription()
+    {
+        static const ColumnsDescription columns_description{
+            { "id",                 std::make_shared<DataTypeUInt128>()                                                                         },
+            { "metric_name",        std::make_shared<DataTypeString>()                                                                          },
+            { "tags",               std::make_shared<DataTypeMap>(std::make_shared<DataTypeString>(), std::make_shared<DataTypeString>())       },
+            { "time_series",        std::make_shared<DataTypeArray>(std::make_shared<DataTypeTuple>(
+                                        DataTypes{std::make_shared<DataTypeDateTime64>(3), std::make_shared<DataTypeFloat64>()},
+                                        Strings{"timestamp", "value"})) },
+        };
+        return columns_description;
+    };
+
+    const ColumnsDescription & getColumnsDescription(TargetTableKind inner_table_kind)
+    {
+        switch (inner_table_kind)
+        {
+            case TargetTableKind::kData:
+            {
+                static const ColumnsDescription columns_description{
+                    { "id",        std::make_shared<DataTypeUInt128>()     },
```

Comment:
UInt128 looks really generous for potentially trillion rows dataset.

16 bytes for ID
8 bytes for Timestamp
8 bytes for Value

Reply:
I'm not sure, maybe UInt64 for ID is enough

Reply:
AFAIK, due to nature of running prometheus collector those id will rotate pretty frequently, and also it does mean, that we are actually not interested in data for particular ID, we interesting in data by particular tag combination and commonly few "id" will match them (`old` and `new` pod for example)

So, i have few ideas in mind: use something like snowflake for id's

Or something more strange, for each "metric name" reserve range of 10000 incremental id's and fill this range with data for that particular metric name and different tags. Just in attempt to get better data locality.

Reply:
Or maybe UInt128 is not too big because this table is always stored ordered by ID first - so the id column will contain sequences of the same values, and that should be compressed pretty well.

> i have few ideas in mind: use something like snowflake for id's

What do you mean?

> Or something more strange, for each "metric name" reserve range of 10000 incremental id's and fill this range with data for that particular metric name and different tags. Just in attempt to get better data locality.

Incremental IDs aren't suitable for implementation because if we use them then `INSERT` to a `TimeSeries` table will require an internal `SELECT` to understand which ID is next. And it would also cause issues with concurrency and when internal tables are replicated. So it's better to use some kind of hash for ID. 

Perhaps ID can be split into two parts: the first 32 or 64 bits would be used for a hash of the metric name, and the remaining 32 or 64 bits would be used for a hash of all the tags' names and values. Then the data for the same metric, but different tags will be stored next to each other.

Reply:
> Or maybe UInt128 is not too big because this table is always stored ordered by ID first - so the id column will contain sequences of the same values, and that should be compressed pretty well.

It's still not good for more data needs to beg decompressed/compressed and also compared (equality condition on that column, also assume that compare of 8 byte field implemented in native way) during queries 

> What do you mean?

Just generate snowflakeID for each "new" tag-value combinations.

>  table would require an internal SELECT to understand which ID is next. 

Well, technically it can be solved by kind of dictionary, in memory which will map hash of tags-values into id's 

> And it would also cause issues with concurrency and when internal tables are replicated.

It's not like we care about consistency of those ID's. we only care about that we can map tag combination to particular  (or list of particular) ID on this node.

Reply:
I decided to make it customizable. The TimeSeries engine supports settings now and also allows to describe columns to change their default types, for example:
```
CREATE TABLE ts(id UInt64) SETTINGS id_algorithm='SipHash_MetricNameLow32_And_TagsHigh32'
```

Reply:
I've implemented something like that but with using the DEFAULT expression instead of the `generatorId` function:

```
CREATE TABLE prometheus
(
        id FixedString(16) DEFAULT murmurHash3_128(metric_name, all_tags),
        metric_name LowCardinality(String),
        all_tags Map(String, String)
) ENGINE=TimeSeries
```